sdk/dotnet/Pulumi/Core/OutputWatcher.cs

```diff
@@ -0,0 +1,33 @@
+// Copyright 2016-2020, Pulumi Corporation
+
+using System.Collections.Immutable;
+using System.Threading.Tasks;
+
+namespace Pulumi
+{
+    /// <summary>
+    /// Allows extracting some internal insights about an instance of
+    /// <see cref="Output{T}"/>. 
+    /// </summary>
+    public static class OutputWatcher
```

Comment:
maybe put into a nested `Utilities` namespace.  `OutputUtilities` might be a better name.

Reply:
I'm not a fan of the word `Utilities` in code... Anything more specific?

Reply:
We could make it an extension method.  My primary goal though is to isolate this so taht people don't get access to this by default.  i.e. they need to do `using Pulumi.Extensions` first.